File: packages/tonic-core/src/hydration/chromaVectorIndex.ts
import { getHydrationChromaCollection } from "./chromaClient";
import type { HydrationRuntimeConfig } from "./runtimeConfig";
import type { VectorFilter, VectorIndex, VectorRecord, VectorSearchResult } from "./vectorIndex";
import type { ChromaCollection } from "./chromaClient";

function toScore(distance: number | undefined): number {
  if (distance == null) {
    return 0;
  }
  return 1 / (1 + distance);
}

export class ChromaVectorIndex implements VectorIndex {
  readonly backend: string;
  readonly collectionName: string;
  private readonly config: HydrationRuntimeConfig;
  private collectionPromise: Promise<ChromaCollection> | null = null;

  constructor(config: HydrationRuntimeConfig) {
    this.config = config;
    this.collectionName = config.collectionName;
    this.backend = "chroma-http";
  }

  private async collection(): Promise<ChromaCollection> {
    if (!this.collectionPromise) {
      this.collectionPromise = getHydrationChromaCollection(this.config);
    }
    return this.collectionPromise;
  }

  async upsert(records: VectorRecord[]): Promise<void> {
    if (records.length === 0) {
      return;
    }
    const collection = await this.collection();
    await collection.upsert({
      ids: records.map((record) => record.id),
      documents: records.map((record) => record.document),
      embeddings: records.map((record) => record.embedding),
      metadatas: records.map((record) => record.metadata ?? {}),
    });
  }

  async similaritySearch(
    queryEmbedding: number[],
    limit: number,
    filter?: VectorFilter,
  ): Promise<VectorSearchResult[]> {
    const collection = await this.collection();
    const result = await collection.query({
      queryEmbeddings: [queryEmbedding],
      nResults: limit,
      where: filter as Record<string, unknown> | undefined,
      include: ["documents", "metadatas", "distances", "embeddings"],
    });
    const ids = result.ids?.[0] ?? [];
    const documents = result.documents?.[0] ?? [];
    const metadatas = result.metadatas?.[0] ?? [];
    const distances = result.distances?.[0] ?? [];
    const embeddings = result.embeddings?.[0] ?? [];
    return ids.map((id, index) => ({
      id,
      document: documents[index] ?? "",
      metadata: (metadatas[index] ?? {}) as Record<string, string | number | boolean | null>,
      embedding: embeddings[index],
      score: toScore(distances[index]),
    }));
  }

  async getByIds(ids: string[]): Promise<VectorSearchResult[]> {
    if (ids.length === 0) {
      return [];
    }
    const collection = await this.collection();
    const result = await collection.get({
      ids,
      include: ["documents", "metadatas", "embeddings"],
    });
    return (result.ids ?? []).map((id, index) => ({
      id,
      document: result.documents?.[index] ?? "",
      metadata: (result.metadatas?.[index] ?? {}) as Record<string, string | number | boolean | null>,
      embedding: result.embeddings?.[index],
      score: 1,
    }));
  }

  async delete(params: { ids?: string[]; filter?: VectorFilter }): Promise<void> {
    const collection = await this.collection();
    await collection.delete({
      ids: params.ids,
      where: params.filter as Record<string, unknown> | undefined,
    });
  }
}

File: merge-tonic-lib/tonic/hydration/chroma_vector_index.py
"""Chroma-backed vector index adapter."""

from __future__ import annotations

from dataclasses import dataclass, field

from .chroma_client import get_hydration_chroma_collection
from .runtime_config import HydrationRuntimeConfig
from .vector_index import VectorFilter, VectorRecord, VectorSearchResult


def _distance_to_score(distance: float | None) -> float:
    if distance is None:
        return 0.0
    return 1.0 / (1.0 + distance)


@dataclass
class ChromaVectorIndex:
    config: HydrationRuntimeConfig
    backend: str = field(init=False)
    collection_name: str = field(init=False)
    _collection: object | None = field(default=None, init=False, repr=False)

    def __post_init__(self) -> None:
        self.collection_name = self.config.collection_name
        self.backend = {
            "http": "chroma-http",
            "persistent": "chroma-persistent",
            "ephemeral": "ephemeral",
        }.get(self.config.mode, f"chroma-{self.config.mode}")

    def _get_collection(self):
        if self._collection is None:
            self._collection = get_hydration_chroma_collection(self.config)
        return self._collection

    def upsert(self, records: list[VectorRecord]) -> None:
        if not records:
            return
        collection = self._get_collection()
        collection.upsert(
            ids=[record.id for record in records],
            documents=[record.document for record in records],
            embeddings=[record.embedding for record in records],
            metadatas=[record.metadata for record in records],
        )

    def similarity_search(
        self,
        query_embedding: list[float],
        limit: int,
        *,
        filter: VectorFilter | None = None,
    ) -> list[VectorSearchResult]:
        collection = self._get_collection()
        result = collection.query(
            query_embeddings=[query_embedding],
            n_results=limit,
            where=filter or None,
            include=["documents", "metadatas", "distances", "embeddings"],
        )
        ids = (result.get("ids") or [[]])[0]
        documents = (result.get("documents") or [[]])[0]
        metadatas = (result.get("metadatas") or [[]])[0]
        distances = (result.get("distances") or [[]])[0]
        embeddings = (result.get("embeddings") or [[]])[0]
        return [
            VectorSearchResult(
                id=id_,
                document=documents[idx] if idx < len(documents) else "",
                metadata=metadatas[idx] if idx < len(metadatas) else {},
                embedding=embeddings[idx] if idx < len(embeddings) else None,
                score=_distance_to_score(distances[idx] if idx < len(distances) else None),
            )
            for idx, id_ in enumerate(ids)
        ]

    def get_by_ids(self, ids: list[str]) -> list[VectorSearchResult]:
        if not ids:
            return []
        collection = self._get_collection()
        result = collection.get(ids=ids, include=["documents", "metadatas", "embeddings"])
        out_ids = result.get("ids") or []
        documents = result.get("documents") or []
        metadatas = result.get("metadatas") or []
        embeddings = result.get("embeddings") or []
        return [
            VectorSearchResult(
                id=id_,
                document=documents[idx] if idx < len(documents) else "",
                metadata=metadatas[idx] if idx < len(metadatas) else {},
                embedding=embeddings[idx] if idx < len(embeddings) else None,
                score=1.0,
            )
            for idx, id_ in enumerate(out_ids)
        ]

    def delete(self, *, ids: list[str] | None = None, filter: VectorFilter | None = None) -> None:
        collection = self._get_collection()
        kwargs: dict[str, object] = {}
        if ids:
            kwargs["ids"] = ids
        if filter:
            kwargs["where"] = filter
        collection.delete(**kwargs)
